### PR TITLE
Move compiler-service.js to TS

### PR DESCRIPTION
**Pull request by RubenRBS**
The main issue with this one is that it needs some type definitions for what a Compiler is, but syncing it across the whole codebase might be a bit difficult. For now, only the bits needed to make this specific module work are added.

### CODE REVIEW

**Comment by RubenRBS** [comment]
Will wait for #3676 to land so I can use the common types defined by that PR before merging

**Comment by jeremy-rifkin** [comment]
#3676 merged

**Comment by RubenRBS** [comment]
Thanks. I've now added the types where necessary

File: static/compiler-service.ts [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+import _ from 'underscore';
+import LRU from 'lru-cache';
+import {EventEmitter} from 'golden-layout';
+
+import {options} from './options';
+
+import {ResultLine} from '../types/resultline/resultline.interfaces';
+
+import jqXHR = JQuery.jqXHR;
+import ErrorTextStatus = JQuery.Ajax.ErrorTextStatus;
```
Comment (matsjla) [inline static/compiler-service.ts:35]: Would love to replace JQuery XHR with fetch in the future but this works for now!

File: static/compiler-service.ts [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+            Sentry.captureException(e);
+        }
+        // TODO: What now? Found no compilers!
+        return {
+            langId: langId,
+            compiler: null,
+        };
+    }
+
+    public getGroupsInUse(langId: string) {
```
Comment (matsjla) [inline static/compiler-service.ts:135]: an explicit return type here would be nice for a complicated chain like this

**Review by matsjla — APPROVED** [review]
Looks great!

File: static/compiler-service.ts (hunk 1 above)
Comment (RubenRBS) [reply static/compiler-service.ts:35]: Hopefuly it becomes usable at some point in the near future 🤞

**Comment by matsjla** [comment]
Sentry deploy 503'd, re-running jobs

**Comment by matsjla** [comment]
weird, lints didn't complain locally :thinking:

**Comment by partouf** [comment]
> weird, lints didn't complain locally thinking

maybe merge main into this?

**Comment by matsjla** [comment]
Could you look at the CI error here @partouf? It's really weird that this shows up on this branch but not on main when this is rebased on main...

**Comment by partouf** [comment]
Some weird things going on here

**Comment by matsjla** [comment]
Think this one is good to go?

**Comment by mattgodbolt** [comment]
Playing catch up here - did we work out what happened here with the CI/Sentry stuff? If not, is there something I can pick up?

Everything look green now - are we cool to merge? @RubenRBS ?

**Comment by matsjla** [comment]
We have a problem with the types for the CMake requests because they don't have a stdout/stderr. I've done some very temporary fixes to get everything working again, but it's something that we should revisit later.

I believe CMake should be working now

**Comment by mattgodbolt** [comment]
Thanks mats

**Pull request merged**